aiida_common_workflows/workflows/relax_and_bands.py

```diff
@@ -0,0 +1,324 @@
+            valid_type=orm.Float,
+            default=lambda: orm.Float(0.0000001),
+            serializer=to_aiida_type,
+            help='The treshold for determining edge cases. Meaning is different depending on bravais lattice.',
+        )
+        spec.input(
+            'seekpath_parameters.with_time_reversal',
+            valid_type=orm.Bool,
+            default=lambda: orm.Bool(True),
+            serializer=to_aiida_type,
+            help='If False, and the group has no inversion symmetry, additional lines are returned.',
+        )
+
+        namspac = spec.inputs.create_port_namespace('relax_inputs')
+        namspac.absorb(CommonRelaxInputGenerator.spec().inputs)
+        namspac['protocol'].non_db = True
+        namspac['spin_type'].non_db = True
+        namspac['relax_type'].non_db = True
+        namspac['electronic_type'].non_db = True
+        namspac['magnetization_per_site'].non_db = True
+        namspac['threshold_forces'].non_db = True
+        namspac['threshold_stress'].non_db = True
+        namspac['engines']['relax']['options'].non_db = True
+
+        namspac2 = spec.inputs.create_port_namespace('bands_inputs')
+        namspac2.absorb(CommonBandsInputGenerator.spec().inputs, exclude=('parent_folder'))
+        namspac2['engines']['bands']['options'].non_db = True
+        namspac2['bands_kpoints'].required = False
+
+        namspac3 = spec.inputs.create_port_namespace('second_relax_inputs')
+        namspac3.absorb(CommonRelaxInputGenerator.spec().inputs, exclude=('structure'))
+        namspac3['protocol'].non_db = True
+        namspac3['spin_type'].non_db = True
+        namspac3['relax_type'].non_db = True
+        namspac3['electronic_type'].non_db = True
+        namspac3['magnetization_per_site'].non_db = True
+        namspac3['threshold_forces'].non_db = True
+        namspac3['threshold_stress'].non_db = True
+        namspac3['engines']['relax']['options'].non_db = True
+        for key in namspac3:
+            namspac3[key].required = False
+            namspac3[key].populate_defaults = False
+            namspac3[key].default = ()
+        namspac3['relax_type'].required = True
+        namspac3['relax_type'].default = RelaxType.NONE
+
+        spec.input('relax_sub_process_class',
+                valid_type=orm.Str,
+                serializer=to_aiida_type,
+                validator=validate_sub_process_class_r
+                )
+        spec.input('bands_sub_process_class',
+                valid_type=orm.Str,
+                serializer=to_aiida_type,
+                validator=validate_sub_process_class_b
+                )
+
+        spec.inputs.validator = validate_inputs
+
+        spec.outline(
+            cls.initialize,
+            cls.run_relax,
+            cls.prepare_bands,
+            if_(cls.should_run_other_scf)(
+                cls.fix_inputs,
+                cls.run_relax
+            ),
+            cls.run_bands,
+            cls.inspect_bands
+        )
```

Comment:
Was wondering if we should have a slightly different outline:
```suggestion
        spec.outline(
            cls.setup,
            if_(cls.should_run_relax)(
                cls.run_relax,
                cls.inspect_relax,
            ),
            if_(cls.should_run_scf)(
                cls.run_seekpath,
                cls.run_scf,
                cls.inspect_scf,
            ),
            cls.run_bands,
            cls.inspect_bands,
            cls.results,
        )
```
I think this is a bit more explicit and clear as to what will happen.

This would make the relax optional, which is still useful, because the user can then compute the bands for an already optimized structure but be sure that the bands are computed in exactly the same way as if it were to include relaxation.

I would also call the second relax run an SCF, because that is what it is, even though we are using the `RelaxWorkflow` but that is merely because we decided that this is the way an SCF is run.

Reply:
I think I am sensing a conceptual difference here. In my mind the user would NOT have the possibility to decide whether  to run an extra scf or not. To determine the presence of an extra step is only the fact that seekpath was used (and in the future maybe the decision to us a different code for bands compared to relax). The user could only select some inputs for this step.

Your idea only gives the additional flexibility to run an extra scf even when it is not necessary (when seekpath is not used) and I believe it is a valid use case. 

However I have strong doubts about your other suggestion to make the relaxation optional
> This would make the relax optional, which is still useful, because the user can then compute the bands for an already optimized structure but be sure that the bands are computed in exactly the same way as if it were to include relaxation.

This can be done already setting the `relax_type` to None in the `relax` input. So it is not needed. I do not see any other case when this  is useful and the price to pay is very high. In fact I believe that making both the `relax`step and the `scf` step optional is difficult to support. How do we understand that the user wants only the `scf` and no the relaxation? If we make `populate_defaults = False` and `required = False` for both relaxation and scf, then we have to implement additional logic that checks that one of the two is set, and that the one selected sets at least the `structure`, `protocol`, `engines`. Basically it makes useless to expose the inputs since None of the properties of the inputs port can be used, they must be implemented all over again. It seems to me an additional complication that has no motivation to exist.

Reply:
I agree that making both optional adds more complexity. But I don't understand your current design. If I understand correctly, in `prepare_bands` you check whether explicit kpoints were specified in the `bands` option, in which case you don't rerun the relax workflow as scf. But the first relax workflow and seekpath are _always_ run. So it is practically guaranteed that the structure changes, even if the user sets `RelaxType.NONE`. Even if the relax doesn't change the structure, it is possible that seekpath still normalizes the structure, which will cause the k-points to longer be commensurate.

I guess what I am saying is that we should make the relax+seekpath optional, or otherwise allowing to directly specify k-points doesn't make any sense. Or we just require relax+seekpath+scf always and just don't support the user specifying their own k-points. At least that would be consistent.

Reply:
Not exactly. The relaxation *always* run. The use of SeeKpath is optional. If you set a kpoints for bands in input, SeekPath is not used. This is a valid use case, for instance, I know the kpoints path along which I want to calculate the bands, and I also want a relaxation of just the atoms coordinates. Of course it does not make much sense to specify a kpoints path and also allow relaxation with variable cell, since you will not know the final cell shape. But the rest is perfectly fine. 

When you DO NOT use Seekpath (i.e. you specify kpoints for bands in input), you do not need an extra scf. In fact, you pass to the bands calculation the `remote_folder` of the relaxation and this guarantees to copy the correct final structure and the density matrix / wave-functions. However, based to your suggestion, I understood that the possibility for the user to trigger an extra scf anyway, is a valid use case. For instance a user wants to do a final scf with "precise" protocol before calculating bands.

To summarize, three options are available
1) relax+seekpath+scf+bands
2) relax+kp_in_input+bands
3) relax+kp_in_input+scf+bands

Reply:
> Of course it does not make much sense to specify a kpoints path and also allow relaxation with variable cell, since you will not know the final cell shape. But the rest is perfectly fine.

This is what was confusing me. Should we maybe have validation in there to prevent this? If you specify kpoints, then the relax cannot change the cell.

> To summarize, three options are available
>
>    relax+seekpath+scf+bands
>    relax+kp_in_input+bands
>    relax+kp_in_input+scf+bands

This is great, this makes things a lot clearer. I would propose to at least describe this in the docstring of the workchain. And then maybe adapt the outline to the following that makes it a lot clearer what is happening:
```python
        spec.outline(
            cls.setup,
            cls.run_relax,
            cls.inspect_relax,
            if_(cls.should_run_seekpath)(
                cls.run_seekpath,
                cls.run_scf,
                cls.inspect_scf,
            ).elif_(cls.should_run_scf)(
                cls.run_scf,
                cls.inspect_scf,
            ),
            cls.run_bands,
            cls.inspect_bands,
            cls.results,
        )
```